### PR TITLE
Implement stuff needed for TurretAI

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/logic/object/update/aiupdate.h [hunk 1]
```diff
@@ -200,21 +242,47 @@ class AIUpdateInterface : public UpdateModule, public AICommandInterface
...
+    unsigned int Get_Next_Mood_Check_Time() const { return m_nextMoodCheckTime; }
+    WhichTurretType Get_Current_Turret() const { return m_currentTurret; }
+    Object *Get_Goal_Object() { return Get_State_Machine()->Get_Goal_Object(); }
+
+    const AIUpdateModuleData *Get_AI_Update_Module_Data() const
+    {
+        return static_cast<const AIUpdateModuleData *>(Get_Module_Data());
+    }
+
+    void Set_Current_Turrt(WhichTurretType t) { m_currentTurret = t; }
```
Comment (tomsons26) [inline src/game/logic/object/update/aiupdate.h]: typo

File: src/game/logic/object/update/aiupdate.h [hunk 2]
```diff
@@ -200,21 +242,47 @@ class AIUpdateInterface : public UpdateModule, public AICommandInterface
...
+    unsigned char unk3[0x20];
     ObjectID m_ignoreObstacleID;
-    unsigned char unk3[0x48];
+    unsigned char unk4[0x48];
     LocomotorSet m_locomotorSet;
     Locomotor *m_curLocomotor;
     LocomotorSetType m_curLocomotorSet;
+    LocoGoalType m_locomotorGoalType;
+    Coord3D m_locomotorGoalData;
+    TurretAI *m_turretAI[2];
```
Comment (tomsons26) [inline src/game/logic/object/update/aiupdate.h]: WhichTurretType should have MAX_TURRETS which then is this 2

File: src/game/logic/object/update/aiupdate.h [hunk 3]
```diff
@@ -41,6 +44,15 @@ enum GuardMode
 {
     GUARD_MODE_UNK
 };
+enum AttitudeType
+{
+    AI_SLEEP = 0xfe,
```
Comment (tomsons26) [inline src/game/logic/object/update/aiupdate.h]: needs to be 32 bit, WB 0043E0B3 shows this

File: src/game/logic/object/update/aiupdate.h [hunk 4]
```diff
@@ -56,21 +68,51 @@ enum LocomotorSetType
     LOCOMOTORSET_COUNT,
 };
 
+enum MoodMatrixAction
+{
+    MM_ACTION_IDLE = 0,
```
Comment (tomsons26) [inline src/game/logic/object/update/aiupdate.h]: 0 isnt needed if thats where we start

File: src/game/logic/object/update/aiupdate.h [hunk 5]
```diff
@@ -56,21 +68,51 @@ enum LocomotorSetType
...
 };
 
 class AIUpdateInterface : public UpdateModule, public AICommandInterface
 {
+    IMPLEMENT_POOL(AIUpdateInterface)
+
 public:
+    enum LocoGoalType
+    {
+        NONE = 0,
```
Comment (tomsons26) [inline src/game/logic/object/update/aiupdate.h]: numbering not needed

File: src/game/logic/ai/turretai.h [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+    TurretStateMachine *m_turretStateMachine;
+    float m_angle;
+    float m_pitch;
+    AudioEventRTS m_turretRotOrPitchSound;
+    unsigned int m_unkFrame1;
+    Team *m_victimInitialTeam;
+    TurretTargetType m_targetType;
+    unsigned int m_unkFrame2;
+    unsigned int m_sleepUntil;
+    unsigned char m_flags;
```
Comment (tomsons26) [inline src/game/logic/ai/turretai.h]: this was likely a bitfield based on xfer

File: src/game/common/statemachine.cpp [hunk 7]
```diff
@@ -707,3 +707,14 @@ void StateMachine::Unlock()
...
+#ifdef GAME_DEBUG_STRUCTS
+Utf8String StateMachine::Get_Current_State_Name()
+{
+    if (m_currentState != nullptr) {
+        return m_currentState->Get_Name();
+    } else {
+        return Utf8String::s_emptyString;
+    }
+}
+#endif
```
Comment (xezon) [inline src/game/common/statemachine.cpp]: Can add EOL here.

File: src/game/logic/ai/aistates.h [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    virtual void On_Exit(StateExitType status) override;
+    virtual StateReturnType Update() override;
+
+    virtual void CRC_Snapshot(Xfer *xfer) override {}
+    virtual void Xfer_Snapshot(Xfer *xfer) override;
+    virtual void Load_Post_Process() override {}
+
+private:
+    NotifyWeaponFiredInterface *m_att;
+};
```
Comment (xezon) [inline src/game/logic/ai/aistates.h]: Can add EOL here

File: src/game/logic/ai/turretai.h [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+    WhichTurretType Get_Which_Turret() const { return m_whichTurret; }
+    float Get_Min_Idle_Scan_Angle() const { return m_data->m_minIdleScanAngle; }
+    float Get_Max_Idle_Scan_Angle() const { return m_data->m_maxIdleScanAngle; }
+    float Get_Recenter_Time() const { return m_data->m_recenterTime; }
+
+    bool Is_Attack_Forced() const { return (m_flags & TURRET_ATTACK_FORCED) != 0; }
+    bool Is_Positive_Sweep() const { return (m_flags & TURRET_POSITIVE_SWEEP) != 0; }
+    bool Is_Flag80() const { return (m_flags & TURRET_80) != 0; }
+
+    void Set_Positive_Sweep(bool set) { m_flags = (TURRET_POSITIVE_SWEEP * set) | m_flags & ~TURRET_POSITIVE_SWEEP; }
```
Comment (xezon) [inline src/game/logic/ai/turretai.h]: Is this a compiler optimization? I have never seen it written this way. Typically one would write it as:

```c++
if (set)
  m_flags |= TURRET_POSITIVE_SWEEP;
else
  m_flags &= ~TURRET_POSITIVE_SWEEP;
```

File: src/game/logic/object/object.cpp [hunk 10]
```diff
@@ -287,3 +287,59 @@ void Object::Set_Captured(bool captured)
         m_privateStatus &= ~STATUS_CAPTURED;
     }
 }
+
+const Weapon *Object::Get_Current_Weapon(WeaponSlotType *wslot) const
```
Comment (xezon) [inline src/game/logic/object/object.cpp:291]: maybe just call it `slot`? `wslot` reads odd.

File: src/game/logic/object/update/aiupdate.h (hunk 1 above)
Comment (jonwil) [reply src/game/logic/object/update/aiupdate.h]: fixed.

File: src/game/logic/object/update/aiupdate.h (hunk 2 above)
Comment (jonwil) [reply src/game/logic/object/update/aiupdate.h]: fixed.

File: src/game/logic/object/update/aiupdate.h (hunk 4 above)
Comment (xezon) [reply src/game/logic/object/update/aiupdate.h]: I think it is fine to be explicit. It also makes it clear that it is not an oversight.

File: src/game/logic/object/update/aiupdate.h (hunk 3 above)
Comment (jonwil) [reply src/game/logic/object/update/aiupdate.h]: fixed.

File: src/game/logic/object/update/aiupdate.h (hunk 4 above)
Comment (jonwil) [reply src/game/logic/object/update/aiupdate.h]: fixed.

File: src/game/logic/object/update/aiupdate.h (hunk 5 above)
Comment (jonwil) [reply src/game/logic/object/update/aiupdate.h]: fixed.

File: src/game/logic/object/update/aiupdate.h (hunk 2 above)
Comment (xezon) [reply src/game/logic/object/update/aiupdate.h]: Also further above at `TurretAIData *m_turretData[2];`

File: src/game/logic/object/update/aiupdate.h [hunk 11]
```diff
@@ -56,21 +68,51 @@ enum LocomotorSetType
...
     static void Parse_Locomotor_Set(INI *ini, void *formal, void *store, const void *user_data);
+
+private:
+    std::map<LocomotorSetType, LocomotorTemplate const *> m_locomotorTemplates;
+    TurretAIData *m_turretData[2];
+    unsigned int m_moodAttackCheckRate;
+    bool m_forbidPlayerCommands;
+    bool m_turretsLinked;
+    int m_autoAcquireEnemiesWhenIdle;
+    friend class AIUpdateInterface;
```
Comment (xezon) [inline src/game/logic/object/update/aiupdate.h:100]: Could you put friends always at the very top of the class body? I think friends are too important to be written so far behind, being barely visible here next to the members.

File: src/game/logic/ai/aistates.h [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+    virtual StateReturnType On_Enter() override;
+    virtual void On_Exit(StateExitType status) override;
+    virtual StateReturnType Update() override;
+
+    virtual void CRC_Snapshot(Xfer *xfer) override {}
+    virtual void Xfer_Snapshot(Xfer *xfer) override;
+    virtual void Load_Post_Process() override {}
+
+private:
+    NotifyWeaponFiredInterface *m_att;
```
Comment (xezon) [inline src/game/logic/ai/aistates.h:72]: Is this supposed to be named `m_att`?

File: src/game/common/statemachine.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/common/statemachine.cpp]: fixed.

File: src/game/logic/ai/aistates.h (hunk 8 above)
Comment (jonwil) [reply src/game/logic/ai/aistates.h]: fixed.

File: src/game/logic/ai/aistates.h (hunk 12 above)
Comment (jonwil) [reply src/game/logic/ai/aistates.h:72]: Yes.

File: src/game/logic/ai/turretai.h (hunk 9 above)
Comment (jonwil) [reply src/game/logic/ai/turretai.h]: It's actually meant to be a bitfield per what toms said. Fixed.

File: src/game/logic/ai/turretai.h (hunk 6 above)
Comment (jonwil) [reply src/game/logic/ai/turretai.h]: fixed.

File: src/game/logic/object/update/aiupdate.h (hunk 2 above)
Comment (jonwil) [reply src/game/logic/object/update/aiupdate.h]: fixed.

**Review by xezon — APPROVED** [review]

**Pull request merged**